[2단계 - 자동차 경주 리팩터링] 짱구(신승철) 미션 제출합니다.

sc0116 opened the pull request:
루피!! 안녕하세요!
1단계 피드백 때 피드백 주신 내용을 다 반영하지 못한 것 같아 2단계에서는 최대한 반영하려고 노력해봤습니다!
이런 귀한 피드백을 다 흡수하지 못해 반성중입니당...

이번에도 몇 가지 궁금한 점들을 적어보았어요!

1. view에서의 getter 사용?
프리코스 때부터 최대한 getter, setter를 사용하지 않으려고 노력중인데요! 2주차 프리코스 피드백에서 `View에서 사용할 데이터라면 getter 메서드를 통해 데이터를 전달한다.` 라는 피드백이 있었지만, 과연 View에서 사용하기 위해 getter를 만드는게 맞나 싶습니다.... View 사용 목적으로 getter를 만들면 public으로 만들어야 하기 때문에 다른 곳에서도 getter를 사용할 가능성이 생겨서 차라리 toString()으로 보내는거에 대해서 어떻게 생각하실지 궁금합니다!

2. stream에서의 getter..?
제 코드에서 우승자를 얻는 메서드는 아래와 같이 구현했습니다.
```java
public List<String> getWinners() {
    return cars.stream()
        .filter(car -> car.isMaxPosition(findMaxPosition()))
        .map(Car::getCarName)
        .collect(Collectors.toList());
}
```

filter 부분에서는 getPosition으로 직접 비교하지 않고, 최대 거리인지 비교하는 메서드를 만들어서 문제를 해결했지만, map은 아무래도 getCarName이 필요해 보입니다. 이 방법이 틀렸나해서 다른 방법을 생각해봐도 결국은 우승자 이름을 담아야 하기 때문에 getter를 사용해야 할 것으로 보이는데요. 이 정도의 사용은 괜찮은건가요? 아니면 혹시 제가 getter를 너무 극단적으로 지양하려고 하는 걸까요...?

---

루피의 피드백을 보고 변수의 사용 범위를 잘 생각하고 선언하자라는 생각이 들었습니다! 어찌보면 사소하고 당연한 부분이지만 제가 이 당연한걸 빠뜨리고 있었는데, 피드백으로 정신이 들게 해주셔서 감사합니다..!!

다시 한번 시간 내주셔서 감사합니다!!😊

sc0116 committed:
refactor(calculator): 상수 접근 제어자 public -> private 변경

sc0116 committed:
refactor(racingcar): 난수 생성 시 시작 범위 설정이 가능하도록 변경

sc0116 committed:
refactor(racingcar): 시도 횟수 RacingGame 생성자에서 검증하도록 변경

sc0116 committed:
refactor(racingcar): 자동차 이름 RacingGame, Car 생성자에서 검증하도록 변경

sc0116 committed:
refactor(racingcar): getter 대신 최대 거리인지 비교하는 메서드 추가

workromancer commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -9,19 +11,23 @@
-    public static final int GAME_END_COUNT = 0;
+    private static final int GAME_END_COUNT = 0;
+
     private final List<Car> cars;
```
일급 컬렉션을 사용할 수 있지 않을까요 ?

일급 컬렉션은 이 글을 참고해보세요~! https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/

sc0116 replied:
미션 중에도 `List<Car> cars` 단 하나를 굳이 `Cars` 클래스로 만들어야 할지 고민을 많이 했었는데 **일급 컬렉션**이란 용어를 찾아보고 나니까 도움이 많이 되었습니다!!

workromancer commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -9,19 +11,23 @@
 public class RacingGame {
 
-    public static final int GAME_END_COUNT = 0;
+    private static final int GAME_END_COUNT = 0;
```
👍

workromancer commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -1,5 +1,7 @@
 package racingcar.model;
 
+import static racingcar.utils.validator.CarNamesValidator.*;
```
해당 클래스에서 사용하는 메서드는 `validateCarName` 하나인 것 같은데 wildcard 로 import 하신 이유가 있으실까요 ?

sc0116 replied:
요거 아마 습관적으로 인텔리제이 자동 코드 포맷기능을 누르는데 이 과정에서 wildcard가 아닌 부분도 wildcard로 바뀌는 것을 오늘 확인해서 설정 들어가서 변경했습니다!

workromancer commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -42,7 +48,7 @@ public void race() {
     public List<String> getWinners() {
         return cars.stream()
-            .filter(car -> car.getPosition() == findMaxPosition())
+            .filter(car -> car.isMaxPosition(findMaxPosition()))
```
👍

workromancer commented on src/main/java/racingcar/utils/validator/CarNamesValidator.java:
```diff
@@ -1,42 +1,37 @@
+    public static void validateDuplicate(final String[] carNames) {
+        Set<String> checkedCarNames = new HashSet<>(Arrays.asList(carNames));
 
-        for (String carName : carNames) {
```
👍

workromancer commented on src/main/java/racingcar/utils/validator/TryCountValidator.java:
```diff
@@ -2,14 +2,23 @@
-    public static void validatePattern(final String inputValue) {
-        if (!inputValue.matches(PATTERN)) {
-            throw new IllegalArgumentException(PATTERN_ERROR_MESSAGE);
+    public static void validateTryCount(final String inputValue) {
```
검증 후에 검증된 값을 반환하는 것은 어떨까요 ?

파라미터로 전달된 값이 검증 후에 사용될지 안될지 확신할 수 없지 않을까요 ?

sc0116 replied:
앗 그렇네요! 검증을 해도 그 후에 만약 검증되지 않은 다른 값을 넣어버린다면 생성시 검증하는 의미가 없어질 수도? 있겠네요!

workromancer replied:
네, 지금은 페어로 미션을 진행하겠지만, 추후에 더 큰 프로젝트를 하게 된다면 다른 개발자가 작성한 코드를 사용할 때도 있고, 짱구가 작성해준 코드를 다른 개발자가 사용할 수도 있을 것 같아요.

위와 같은 구조라면 다른 개발자가 사용할 때, 검증되지 않은 값으로 객체를 생성하는 경우가 생길 수도 있지 않을까요 ㅎㅎ

workromancer commented on src/main/java/racingcar/utils/validator/TryCountValidator.java:
```diff
@@ -2,14 +2,23 @@
 
 public class TryCountValidator {
 
-    private static final String PATTERN = "^([1-9])([0-9])*$";
```
정규표현식 제거 👍

workromancer commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -20,23 +17,17 @@ public InputView() {
+        return input.nextLine().split(CAR_NAMES_SPLIT_REGEX);
     }
 
     private void printCarNameInputMessage() {
```
이 메서드는 콘솔에 출력하는 역할로 보이는데 InputView 클래스의 메서드로 분류하신 이유가 있으실까요 ?

sc0116 replied:
처음에 입출력을 구분할 때 정말 입력, 출력으로 구분하지 않고 출력이어도 입력을 하라는 안내 문구?(`System.out.println(CAR_NAME_INPUT_MESSAGE`)들은 InputView에 넣어두다가 단순히 메서드(`printCarNameInputMessage()`)로 분리시켰던 건데 혹시 입력을 하라는 안내문구 같은것도 출력으로 봐야할까요???

workromancer replied:
프로그래밍에는 정답은 없어서요 ㅎㅎ 단순히 어떤 의도로 작성하신지 궁금해서 여쭤봤습니다 !

workromancer commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -20,23 +17,17 @@ public InputView() {
-        TryCountValidator.validatePattern(tryCount);
-
-        return Integer.parseInt(tryCount);
+        return input.nextLine();
```
👍

workromancer reviewed (changes requested):
짱구~ 안녕하세요.

step1 에서 드린 피드백 고민 주신 것 확인하였습니다~! 
제가 지난 코멘트를 너무 늦게 확인해서 짱구도 혹시 놓치실까봐 답변과 함께 링크 전달드릴께요 !

1. https://github.com/woowacourse/java-racingcar/pull/312#discussion_r805934327
2. https://github.com/woowacourse/java-racingcar/pull/312#discussion_r808854112

추가로 이번에 리뷰 요청 주신 부분에 대해서도 코멘트를 남겨두었습니다. 확인 후에 고민해보시고 다음 리뷰 요청 때, 궁금하신 점 코멘트나 DM 주세요 !

이번에 질문 주신 두 내용에 대한 답변도 추가합니다 !

Q. view에서의 getter 사용?
A.
View 에서 toString 보다 Getter 를 사용하는 것은 괜찮은 것 같아요. Getter 를 지양해야 할 곳은 이번에 리팩토링 해주신 [RacingGameL45](https://github.com/woowacourse/java-racingcar/pull/370/files#diff-35605c700c29bbe88da04bcd83b40b2523753cfde13d162cd3192baee2a66068L45) 같은 부분이 아닐까요 ?!

Q. stream 에서 getter
A.
이 부분도 고민이 많이 되실 것 같아요. 승자를 반환할 때, 이름을 바로 반환하지 않고 승리한 Car 를 반환해보는 것은 어떨까요 ?

sc0116 committed:
refactor(racingcar): 입출력 View 리팩터링

- View 정적 클래스로 변경
- OutputView 우승자 목록 생성 메서드의 파라미터로 Car 자체를 가져오도록 변경
- split할 때 마지막이 null인 경우도 포함하도록 limit 인수 추가

sc0116 committed:
refactor(racingcar): 검증 후 값 반환하도록 변경

sc0116 committed:
refactor(racingcar): 일급 컬렉션 사용, Model에서 getter 사용하지 않도록 변경

workromancer commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,45 @@
+
+import racingcar.utils.RandomNumber;
+
+public class Cars {
```
일급컬렉션을 구현하면 단순한 컬렉션을 사용할 때보다 역할과 책임이 명확한 컬렉션을 만들 수 있겠지요 ? ㅎㅎ

일급컬렉션에 대하여 고민하신것이 느껴지네요 ㅎㅎ 👍

workromancer commented on src/test/java/racingcar/validator/CarNamesValidatorTest.java:
```diff
@@ -1,19 +1,19 @@
     void 자동차_이름_중복() {
         String[] carNames = {"aa", "bb", "aa"};
-        assertThatThrownBy(() -> CarNamesValidator.validateDuplicate(carNames))
+        assertThatThrownBy(() -> new Cars(carNames))
```
Cars 를 테스트하는 테스트 클래스를 만들어도 좋지 않을까요 ?

workromancer commented on src/test/java/racingcar/validator/TryCountValidatorTest.java:
```diff
@@ -1,6 +1,6 @@
 package racingcar.validator;
 
-import static org.assertj.core.api.Assertions.*;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
```
wildcard import 보다 구체적인 import 문을 사용하면 import 문을 통해서 해당 클래스에서 어떤 것들을 사용할 수 있는지 확인하기 훨씬 용이해지겠지요 ? 👍

workromancer commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -24,13 +23,18 @@ public int getPosition() {
 
-    public boolean isMaxPosition(int max) {
-        return position == max;
+    public boolean isMaxPosition(Car car) {
```
👍

workromancer reviewed (approved):
짱구 안녕하세요~

첫번째 미션 고생 많으셨습니다.
`List<Car>` 를 일급 컬렉션으로 리팩토링하여 적절한 역할과 책임을 부여하신 것이 인상적이었어요. 👍

코멘트 몇개 남겨두었는데, 다음 미션 전에 확인 부탁드려요.

이번 미션은 이만 머지하도록 할께요.
이번 미션을 되돌아보면서 다음 미션을 준비해보시는 것이 어떨까요 ?

추가로 궁금하신 부분 있으시면, 언제든 DM 주세요.

고생하셨습니다. 다음 미션도 화이팅이예요 💪
